jupyter_client/kernelspec.py

```diff
@@ -54,19 +56,47 @@ def to_json(self):
         """
         return json.dumps(self.to_dict())
 
+
+_kernel_name_pat = re.compile(r'^[a-z0-9._\-]+$', re.IGNORECASE)
+
+def _valid_kernel_name(name):
+    """Check that a kernel name is valid.
+    
+    Returns True if valid, False if not.
+    """
+    # quote is not unicode-safe on Python 2
+    return _kernel_name_pat.match(name)
+
+
+_kernel_name_description = "Kernel names can only contain ASCII letters and numbers and these separators: -._"
+
+
 def _is_kernel_dir(path):
     """Is ``path`` a kernel directory?"""
     return os.path.isdir(path) and os.path.isfile(pjoin(path, 'kernel.json'))
 
+
 def _list_kernels_in(dir):
     """Return a mapping of kernel names to resource directories from dir.
 
     If dir is None or does not exist, returns an empty dict.
     """
     if dir is None or not os.path.isdir(dir):
         return {}
-    return {f.lower(): pjoin(dir, f) for f in os.listdir(dir)
-                        if _is_kernel_dir(pjoin(dir, f))}
+    kernels = {}
+    for f in os.listdir(dir):
+        path = pjoin(dir, f)
+        if not _is_kernel_dir(path):
+            continue
+        key = f.lower()
+        if not _valid_kernel_name(key):
+            warnings.warn("Invalid kernelspec directory name (%s): %s",
+                _kernel_name_description, path, stacklevel=3,
+            )
+            continue
```

Comment:
If we want to have a grace period for invalid names, we need to eliminate this `continue` since it will just skip over this `key`